jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```
@@ -804,32 +803,78 @@ public void execute(Runnable job)
-                continue;
+            // idle can become negative, which means there are queued tasks.
+            if (_counts.compareAndSet(counts, threads + (startThread ? 1 : 0), idle + (startThread ? 0 : -1)))
+                break;
+        }
 
-            break;
+        // We should always be able to add the job to the queue as we checked the capacity above,
+        // However, a starting thread may not yet have polled the queue, so this could fail in a race.
+        // Return if we are able to queue the job
+        if (_jobs.offer(job))
+        {
+            // Start additional threads
+            if (startThread)
+                startThread();
+            return;
         }
 
-        if (!_jobs.offer(job))
+        // Start additional threads
+        if (startThread)
+            startThread();
+
+        // We lost the race with a starting thread, so we can do a blocking put that will wait for the starting thread
+        if (_putSupported)
         {
-            // reverse our changes to _counts.
-            if (addCounts(-startThread, 1 - startThread))
-                LOG.warn("{} rejected {}", this, job);
-            throw new RejectedExecutionException(job.toString());
+            try
+            {
+                _jobs.put(job);
+                return;
+            }
+            catch (InterruptedException e)
+            {
+                addCounts(0, 1);
+                Thread.currentThread().interrupt();
+                throw new RejectedExecutionException(e);
+            }
+            catch (UnsupportedOperationException uoe)
+            {
+                LOG.warn("Detected bounded thread pool queue {} without put support", _jobs.getClass());
+                _putSupported = false;
+            }
         }
 
-        if (LOG.isDebugEnabled())
-            LOG.debug("queue {} startThread={}", job, startThread);
+        // Put was not supported, so we will just spin instead
```

I don't think this is a good idea.
We have already been reported in the past about excessive CPU consumption due to spin.
We tried `put()`, now we reject, no spinning.

---

Hmmm maybe. We do warn and its not a bad spin and I do throttle.     So I'll leave it in for now...   Perhaps we need to see how many Q implementations do not have put?

---

I'd really take it out.
If someone complains, then we can add it back, but it's the rare case of a rare case that will bite us back.
I'd prefer fail fast and reject.
`parkNanos()` will also spin, so now we are double-spinning, so I'd remove the spin.